Add CircleCI Deploy Option

vsoch opened the pull request:
This PR will address #350 to add a circle configuration (meaning a hidden folder, `.circleci` with a `config.yml` file) to do the following:

 - build a container from a user specified repository with notebook
 - deploy to Docker Hub

The base could additionally be customized by the user to do more testing or extraction of metadata, and this should be considered a base or skeleton. To help with this, I have also made the following changes to documentation:

 - I added a deploy.rst page, rendered from the index.rst. I usually do markdown so please yell at me if I mucked up some formatting :)
 - I replaced the (broken) Data Science Notebook repository with the other example that is working.

What you might want to do is to connect this repo to CircleCI and then define the required environment variables and create the Docker Hub repository (read the deploy.rst instructions) so that you can build and deploy an example to go with this repository. if you wish, you can also add a filter to only deploy on master merge!

For more verbose description, please see the issue here ! https://github.com/jupyter/repo2docker/issues/350#issuecomment-405194344

betatim commented:
Short thought: should we move `.circleci` to somewhere else (maybe `examples/`?) to make it clear that this is an example of how you would setup your repository if you wanted to use `repo2docker` to continuously build the image for your repo?

For repo2docker we don't really want to build a docker image all the time and might want to use circleci for other things. At which point it would become confusing which parts of the circle CI config are for repo2docker testing and which are about setting up this continuous build thing?

betatim commented on .circleci/config.yml:
```diff
@@ -0,0 +1,171 @@
+            echo "Container name set to ${CONTAINER_NAME}"
+            if [[ -n "$DOCKER_PASS" ]]; then
+                  docker login -u $DOCKER_USER -p $DOCKER_PASS
+                  docker push ${CONTAINER_NAME}:latest
```
I'd push using the first seven (or so) digits of the commit hash as tag instead of `latest`. Get people in the habit of explicitly specifying version :)

betatim replied:
We should document somewhere (even if the `.rst` docs already do so) here or when CONTAINER_NAME is defined that you have to own the docker hub organisation that matches your GitHub organisation for this setup to work. At least I think that is true.

vsoch replied:
Haha, sure thing! How about if the user explicitly defines a `DOCKER_TAG` we use that, but if not, add the commit tag instead? It looks like I can grab the commit with `CIRCLE_SHA1`

betatim replied:
I'd use the SHA1 as default. If people hate it they can redefine that bit of circleci config. I think that dockerhub always has a `latest` that points to the latest image.

vsoch replied:
okay just pushed a change for `DOCKER_TAG`, let's see if it works!

betatim commented on .circleci/config.yml:
```diff
@@ -0,0 +1,171 @@
+            fi
+            echo "Saving ${CONTAINER_NAME} to container.tar.gz"
+            mkdir -p /tmp/cache
+            docker save ${CONTAINER_NAME}:latest \
```
Will this save the state after running the command on L77 or state before? Both could be interesting to have but I'd add a comment to clarify which is going to be save here.

vsoch replied:
What is L77?

vsoch replied:
I know what this is now :)

vsoch replied:
I thought you were talking about some derivation of Fortran 77 (F77 -> L77) and I just had no idea.

betatim commented:
This is awesome! Some short(ish) thoughts below:

Should we move `.circleci` to somewhere else (maybe `examples/`?) to make it clear that this is an example of how you would setup your repository if you wanted to use `repo2docker` to continuously build the image for your repo?

Is this the simplest possible circleCI config we can cook up? For me the level of complexity is Ok, but I am thinking of someone who is new to this whole container and CI thing. Every extra line that does something which is "nice to have" but not really required adds complexity and potential for cargo cult'ing. If there are steps we can remove I'd maybe make a "basic" version and a "doing it really well".

For the repo2docker repository we don't really want to build a docker image all the time and might want to use circleci for other things. At which point it would become confusing which parts of the circle CI config are for repo2docker testing and which are about setting up this continuous build thing?

vsoch commented:
With the changes that are coming to circle (orbs) you will actually be able to maintain separate files, so I don't think it would be confusing. But it's up to you! I'm not a jupyter organization person so I can't make a new repository with a working example, which I think is essential. I would suggest that it doesn't hurt now to have this (because it's useful and needed) and either decide to keep here (because it's not confusing when orbs are released) or when you **do** set up circle here (and decide you don't like having both) to create the dummy repo. It's definitely up to you!

betatim commented:
> 1. Multiple Workflows

I think of myself as a proficient Circle CI and travis user, and yet I have to grab the docs all the time when I do reconfigure one of them. So I was wondering if we could have a directory that I can copy into my existing repository that "does the right thing" without having to study the example here to remove the bits which are for testing repo2docker and aren't needed for my repo.

> 2. A New Separate Repository

I think this is a good idea. We could place it in the https://github.com/binder-examples/ organisation where we already have quite a few binder related example repositories. We could configure a cron job for that repo so that we get builds once a day/week so we notice when a change to repo2docker breaks this setup. (Cron builds would be a cool advanced feature anyway ;) )

I would want people to take the example config from somewhere and add it to their existing repository. Not start from cloning repo2docker and modifying it to add their work.

vsoch commented on .circleci/config.yml:
```diff
@@ -0,0 +1,171 @@
+            echo "jupyter-repo2docker --debug --user-name username --user-id 1000 --no-run --image-name ${CONTAINER_NAME} ${REPO_NAME}"
+            jupyter-repo2docker --debug --user-name username --user-id 1000 --no-run --image-name "${CONTAINER_NAME}" "${REPO_NAME}"
+            echo "3. Running without server (example)"
+            docker run -td --name repo2docker --entrypoint /bin/bash "${CONTAINER_NAME}"
```
We don't actually need to run, that was there as an example (and I just removed per your comment to make it streamlined!)

Ohhh L77 --> line 77! I'm dense, hehe. Yes! I totally agree on having notes / docs for how to expand this, do you want me to add these lines back in even though they aren't streamlined? I think when the orbs come out we could provide several templates (for doing different things with the notebooks in the cI) and can develop them as the need arises. This PR is not meant to address these additional helpers / docs - it's scoped to be the basic skeleton of this step.

vsoch commented:
@betatim with the changes I made (removing the run) there is no extra work needed other than defining credentials in the CircleCI web interface. The user clones, connects, and defines those variables, that's as simple as I can get it! (no needing to edit the config file, I agree that is always challenging.)

Sure that works! I'm not sure how I can help if the repository doesn't exist though... ideas?

vsoch commented:
Okay `DOCKER_TAG` looks good! See the updated build here:
https://circleci.com/gh/vsoch/repo2docker/22

And the tag is now very specific:
https://hub.docker.com/r/vanessa/repo2docker/tags/

This new strategy ensures that every single deployed image is maintained on Docker Hub (as opposed to having many "latests" with different sha sums) so it's a neat idea.

For next steps - if you want to make me a repository to issue PR to where an example (connected to circleci) could be depployed, I'd be happy to remove the .circleci folder from here, and put it over there. The docs can still be added for deploy.rst here, but they will point the user to the other respository. The other repository can be exclusively for many examples of different tests/ other one might use for a circleci deployment, and we can add them as users request/ need.

betatim commented:
> For next steps - if you want to make me a repository to issue PR to where an example (connected to circleci) could be depployed, I'd be happy to remove the .circleci folder from here, and put it over there. The docs can still be added for deploy.rst here, but they will point the user to the other respository. The other repository can be exclusively for many examples of different tests/ other one might use for a circleci deployment, and we can add them as users request/ need.

I think that is the option I'd champion. Let's see if any of the others have comments. If we decide to do this I'll create a repo in binder-examples.

vsoch commented:
woohoo! I'm down!

vsoch commented:
Let's leave this one open - I'll remove the config.yml from here, and update the docs to be referencing the binder-examples. That way, the user can still find what he/she is looking for even if the files aren't hosted alongside.

minrk commented:
I think the proposal you folks came to sounds like a great way to go.

betatim commented:
I created https://github.com/binder-examples/continuous-build. The name is not perfect, suggestions welcome.

👍  to keeping docs here in this PR and the CircleCI config in https://github.com/binder-examples/continuous-build

vsoch commented:
on it!

vsoch commented:
@betatim I've updated this pull request in the following ways:

 - the entire circle config folder is removed, and the jupyter team can add something back when desired to write tests for repo2docker (I agree this is the more appropriate approach to take!)
 - I have updated the deploy.rst docs to now reference the new "continuous-build" folder.
 - I don't think the name is bad at all, it says exactly what it is and I like it.

I have also opened a pull request to the other repository that will include the "default" configuration example, along with (in the future) configurations with other bells and whistles.

 - The [pull request is here](https://github.com/binder-examples/continuous-build/pull/1)
 - The (same) deployment as before from the new repo is [here](https://circleci.com/workflow-run/10042199-d8aa-4e41-9764-5535c812fd87).

It's pretty cool that I didn't need to redefine the environment variables, I could simply import them from my previous project! This is really cool for organizations because it means that (hidden credentials) to deployment places could be shared with others in the organization without actually sharing them. The last few letters of variables ARE shown however, so this should be done with care (if at all). If you would like to add me as a contributor to the other binder-examples repo, I'd be glad to monitor and help users with custom / new configurations. We will have a little recipe book of templates!

vsoch commented:
Update for those following this PR - the repo with just the recipes is now merged and ready for use, and we can update the docs here (per this PR) and then (if jupyter has other forms of communication / docs) share with users, and start to brainstorm how to capture metadata, test, etc.

jzf2101 commented:
I agree, creating a separate issue would be a good idea.

vsoch commented:
@jzf2101 see https://github.com/binder-examples/continuous-build/ ! We have one more PR in to set a default `REPO_NAME` so the circle build doesn't fail given that the user forgets to define one. I had originally imagined we would want it to fail, but probably not.

vsoch commented:
oh yes that's a great idea! +1 - maybe put this as an issue on the other repo (unless you already did, haha).  I think it could work both ways actually, given that there is a travis and circle builder template, for example, the files are isolated (e.g. travis.yml vs .circleci so you would want to have certainty of testing from the same base (argument for keeping one repo). On the other hand, let's say we have a bunch of just circle templates, it would be more clear to have separate repos for each (and just take care to test them separately as appropriate).

betatim commented:
Let's move the discussion of naming and whether or not to have multiple CIs in one repo to https://github.com/binder-examples/continuous-build/issues/5

betatim commented on docs/source/deploy.rst:
```diff
@@ -0,0 +1,127 @@
+2. build
+3. push to Docker Hub
+
+Also provided (or will be provided as they are requested) are other configuration templates in
```
I'd change it to say what it does provide and avoid speculating about the future.

vsoch replied:
```python
from __future__ import vanessa
```

vsoch commented on docs/source/deploy.rst:
```diff
@@ -0,0 +1,127 @@
+.. _usage:
+
+Deploy ``repo2docker`` 
```
sure thing!

vsoch replied:
I usually think of deploy as == push, how do you usually define / use it?

vsoch commented on docs/source/deploy.rst:
```diff
@@ -0,0 +1,127 @@
+
+You should then be able to pull your new container, and run it! Here is an example:
+
+  docker pull vanessa/repo2docker
```
yep!

vsoch commented:
Did you need any more clarification on my part? We have this documentation linked from the continuous build repo, so we should get deployed soon!

betatim commented:
Nope, just slow because it was the weekend :) Merged!

vsoch commented:
Looks great! https://repo2docker.readthedocs.io/en/latest/deploy.html  and Happy Monday :)

vsoch commented:
Thanks again for all the work and review here! Please ping me if I can help with anything in the future.
